primeqa/mrc/processors/preprocessors/tydiboolqa_bpes.py

```diff
@@ -0,0 +1,86 @@
+import itertools
+from operator import itemgetter
+from operator import sub
+from typing import Optional
+from typing import List, Iterable, Tuple, Any, Dict, Union
+
+from datasets.arrow_dataset import Batch
+from transformers import BatchEncoding
+
+from datasets import Dataset
+from datasets.features.features import Sequence, Value
+
+from primeqa.mrc.processors.preprocessors.tydiqa import TyDiQAPreprocessor
+from primeqa.mrc.data_models.target_type import TargetType
+
+
+# BPES preprocessor
+
+class TyDiBoolQAPreprocessor(TyDiQAPreprocessor):  # TODO type signatures for all methods
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+
+
+
+    def _make_passage_spans(self, examples, features_dict):
+        target_type=features_dict['target_type']
+        example_idx=features_dict['example_idx']
+        offsets=features_dict['offset_mapping']
+        if target_type == TargetType.NO or target_type == TargetType.YES:
+            start_position = features_dict['start_positions']
+            end_position = features_dict['end_positions']
+            input_ids = features_dict['input_ids']
+            example = examples[example_idx]
+            target=example['target']
+            passage_index=target['passage_indices'][0]
+            passage_start_offset=example['passage_candidates']['start_positions'][passage_index]
+            passage_end_offset=example['passage_candidates']['end_positions'][passage_index]
+
+            token_is_inside_passage = [passage_start_offset <= tok_start and
+                                        tok_end < passage_end_offset
+                                        for (tok_start, tok_end) in offsets]
+
+            # start/end in tokens is given by the first True and last True in after the query in token_is_inside_passage
+            query_end_position=input_ids.index(self._tokenizer.sep_token_id)+2
+            try:
+                passage_start_position = token_is_inside_passage.index(True,query_end_position)
+            except ValueError as x:
+                passage_start_position = query_end_position
+            try:
+                passage_end_position = token_is_inside_passage.index(False, passage_start_position)
+            except ValueError as x:
+                passage_end_position = len(token_is_inside_passage)-1
+
+#            print(f'{target_type} {passage_start_position} {start_position} {end_position} {passage_end_position}')
+#            if target_type == TargetType.SPAN_ANSWER:
+#                assert( passage_start_position <= start_position <= end_position <= passage_end_position)
+            features_dict['start_positions'] = passage_start_position
+            features_dict['end_positions'] = passage_end_position
+        # always return features_dict
+        return features_dict
+
+
+    def process_train(self, examples: Dataset) -> Tuple[Dataset, Dataset]:
+        """ 
+sample comparison of features and updated_features on Tydi train
```

Comment:
I don't know if this example is necessary. Perhaps an explanation would be more useful?

Reply:
I strong feel that the example will be valuable if I need to debug this again